{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.0.11'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "import deeplake\n",
    "\n",
    "deeplake.__version__\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: BUGGER_OFF=True\n"
     ]
    }
   ],
   "source": [
    "%env BUGGER_OFF=True\n",
    "\n",
    "deeplake.client.config.USE_LOCAL_HOST = True\n",
    "\n",
    "LOCAL_TOKEN = \"eyJhbGciOiJIUzUxMiIsImlhdCI6MTY2MTI0OTk1OCwiZXhwIjo0ODE0ODQ5OTU4fQ.eyJpZCI6InByb2dlcmRhdiJ9.tL69TRCFU2v8mElx3UjGauzdjgHrYIYT37PKNIPr_WqVUdJ4zFt1wYrmBdJw7uD9u5ueMzfUYP0EcU1YMMV4HQ\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s3://platform-test/mnist-train loaded successfully.\n"
     ]
    }
   ],
   "source": [
    "DEFAULT_CREDS = {\n",
    "    \"aws_access_key_id\": \"UOAYBIDVZ0NLDT5JVJ7T\",\n",
    "    \"aws_secret_access_key\": \"ez49zQ3W9Cj7GV1vs7F8nz3ugeLS53qw25qYOeUF\",\n",
    "    \"aws_session_token\": \"\",\n",
    "    \"endpoint_url\": f\"https://s3.us-east-1.wasabisys.com\",\n",
    "    \"region\": \"us-east-1\",\n",
    "}\n",
    "path = \"s3://platform-test/mnist-train\"\n",
    "\n",
    "ds = deeplake.load(path, creds=DEFAULT_CREDS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s3://hub-2.0-datasets-n/fashionpedia_small/ loaded successfully.\n"
     ]
    }
   ],
   "source": [
    "creds = {\n",
    "    \"aws_access_key_id\": \"ASIAYLX7F2O7MMF5N7EM\",\n",
    "    \"aws_secret_access_key\": \"jyI7KsvluId2c9oEquoLiCZX9k5RQUhkN1RQWFp0\",\n",
    "    \"aws_session_token\": \"IQoJb3JpZ2luX2VjEGgaCXVzLWVhc3QtMSJHMEUCIQD9qko/whKDUSUdByyHxxG6O+XQvJIkLcatYgJvu8AbbwIga5QQHPRSJBd0niFb9KNWP7ZvVaJ0slBZgelWstbYDuMqrAMIQRAAGgw1NzQ5ODcwMzE0ODYiDEdMv5l4EYYaLtcEkiqJA5C9AlWVkN0kOnHGb3A2bTTsztqanSmqwFxeO4gkA+L9L+jhezYayolDzG59KqVrXzl6BygwPRXjROP6kLJ3xPHsSjIEStlR5cLzSqiPJUWOh3lHw69wqXWdVw4gsXeKs2rPf5NmeMaOuPJlFvPAqJwJuOdNB6lNT9Xq9fNujclnjAZV0GUbN12tNRJw8Y75qE2SbC6mrm8nf00XkI+rXKdKPG+3zWokHwuy+YXqSYnVa2yKVJcsaSSMXdChbRyrRDLVFQ9ZDk80VdsEVWqCrUjptxyF4x3L6xUzLMv8f30Yyl1fFlNP1fTJBNC19YXBY4s6L5DVKPa2O+3lNCEcPQjcpEcxM17mNbtBsqQtiRPaV9qmgDB9D+mwa2HBrhcW+Kw3verC+ZnUs+CSa6phQlbBZ8JOZUmx41Bjz0ng13iFwpm3nCmkTMpU/jOXMLmKP2CGCAbei00/Jy82nzp3ApoWlicYPj9q0PQ0Pw5J4o+RioTDhUx6wslvDIVQVqr2Pxn5+sPiMG78AzDo1r6aBjqmAXqJRj44WZVNAY+9UGfRb0B86z/as7WTAHlgL3dML6wvGgAI1xtudfj04ihNnfK3FL4QxaIoqxun/K9v/tJ11LCeJZdfhr91EK7urY6yY7lf1IutUY8K5ederbBFcMTAfUqDXPot+WXIt89kGgaCYIEZus61emvwITYlNJUt2J+Lb5X22Jk/rXOshVt7+NGt7P7c6oio9u51mhfgsIw0iwpTtU/dlOA=\",\n",
    "    \"region\": \"us-east-1\",\n",
    "}\n",
    "\n",
    "ds = deeplake.load(\"s3://hub-2.0-datasets-n/fashionpedia_small/\", creds=creds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "InvalidSourcePathError",
     "evalue": "Source dataset is already a Deep Lake Cloud dataset.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidSourcePathError\u001b[0m                    Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m connected_ds \u001b[39m=\u001b[39m deeplake\u001b[39m.\u001b[39mconnect(token\u001b[39m=\u001b[39mLOCAL_TOKEN, src_path\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mhub://hub-2.0-datasets-n/fashionpedia_small/\u001b[39m\u001b[39m\"\u001b[39m, creds_key\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCONNECT_CREDS2\u001b[39m\u001b[39m\"\u001b[39m, dest_path\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mhub://progerdav/connect_api\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/projects/activeloop/deeplake/deeplake/api/dataset.py:933\u001b[0m, in \u001b[0;36mdataset.connect\u001b[0;34m(src_path, creds_key, dest_path, org_id, ds_name, token)\u001b[0m\n\u001b[1;32m    915\u001b[0m \u001b[39m@staticmethod\u001b[39m\n\u001b[1;32m    916\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\n\u001b[1;32m    917\u001b[0m     src_path: \u001b[39mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    922\u001b[0m     token: Optional[\u001b[39mstr\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    923\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Dataset:\n\u001b[1;32m    924\u001b[0m     \u001b[39m\"\"\"Connects a dataset at the specified path to Deep Lake.\u001b[39;00m\n\u001b[1;32m    925\u001b[0m \n\u001b[1;32m    926\u001b[0m \u001b[39m    Args:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    931\u001b[0m \u001b[39m        Dataset: The connected object.\u001b[39;00m\n\u001b[1;32m    932\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 933\u001b[0m     connected_id \u001b[39m=\u001b[39m connect_dataset_entry(\n\u001b[1;32m    934\u001b[0m         src_path\u001b[39m=\u001b[39;49msrc_path,\n\u001b[1;32m    935\u001b[0m         creds_key\u001b[39m=\u001b[39;49mcreds_key,\n\u001b[1;32m    936\u001b[0m         dest_path\u001b[39m=\u001b[39;49mdest_path,\n\u001b[1;32m    937\u001b[0m         org_id\u001b[39m=\u001b[39;49morg_id,\n\u001b[1;32m    938\u001b[0m         ds_name\u001b[39m=\u001b[39;49mds_name,\n\u001b[1;32m    939\u001b[0m         token\u001b[39m=\u001b[39;49mtoken\n\u001b[1;32m    940\u001b[0m     )\n\u001b[1;32m    942\u001b[0m     \u001b[39mreturn\u001b[39;00m dataset\u001b[39m.\u001b[39mload(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mhub://\u001b[39m\u001b[39m{\u001b[39;00mconnected_id\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m, token\u001b[39m=\u001b[39mtoken, verbose\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/projects/activeloop/deeplake/deeplake/util/connect_dataset.py:16\u001b[0m, in \u001b[0;36mconnect_dataset_entry\u001b[0;34m(src_path, creds_key, dest_path, org_id, ds_name, token)\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect_dataset_entry\u001b[39m(\n\u001b[1;32m      8\u001b[0m     src_path: \u001b[39mstr\u001b[39m,\n\u001b[1;32m      9\u001b[0m     creds_key: \u001b[39mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     13\u001b[0m     token: Optional[\u001b[39mstr\u001b[39m],\n\u001b[1;32m     14\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mstr\u001b[39m:\n\u001b[1;32m     15\u001b[0m     \u001b[39mif\u001b[39;00m is_hub_cloud_path(src_path):\n\u001b[0;32m---> 16\u001b[0m         \u001b[39mraise\u001b[39;00m InvalidSourcePathError(\n\u001b[1;32m     17\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mSource dataset is already a Deep Lake Cloud dataset.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     18\u001b[0m         )\n\u001b[1;32m     20\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_path_connectable(src_path):\n\u001b[1;32m     21\u001b[0m         \u001b[39mraise\u001b[39;00m InvalidSourcePathError(\n\u001b[1;32m     22\u001b[0m             \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mSource path may only be an s3 or gcs path. Got \u001b[39m\u001b[39m{\u001b[39;00msrc_path\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m     23\u001b[0m         )\n",
      "\u001b[0;31mInvalidSourcePathError\u001b[0m: Source dataset is already a Deep Lake Cloud dataset."
     ]
    }
   ],
   "source": [
    "connected_ds = deeplake.connect(token=LOCAL_TOKEN, src_path=\"s3://hub-2.0-datasets-n/fashionpedia_small/\", creds_key=\"CONNECT_CREDS2\", dest_path=\"hub://progerdav/connect_api\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from deeplake.core.dataset.deeplake_cloud_dataset import DeepLakeCloudDataset\n",
    "\n",
    "\n",
    "isinstance(ds, DeepLakeCloudDataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "class A:\n",
    "    def __init__(self) -> None:\n",
    "        self.a = \"AAA\"\n",
    "        self.b = None\n",
    "\n",
    "    def method(self):\n",
    "        print(self.a)\n",
    "\n",
    "    def convert(self):\n",
    "        self.__class__ = B\n",
    "        self.__init__()\n",
    "\n",
    "\n",
    "class B(A):\n",
    "    def __init__(self) -> None:\n",
    "        self.a = \"AAAA\"\n",
    "        self.b = \"BBBB\"\n",
    "\n",
    "    def method(self):\n",
    "        print(self.a + \" \" + self.b)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAA BBBB\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 4 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 5 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 1 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 2 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Encountered connection error, retry 3 out of 5\n",
      "  warnings.warn(*args, **kwargs)\n",
      "/Users/davitgyulnazaryan/projects/activeloop/deeplake/deeplake/util/warnings.py:7: UserWarning: Connection re-established after 3 retries.\n",
      "  warnings.warn(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "a = A()\n",
    "\n",
    "a.convert()\n",
    "\n",
    "a.method()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "dc1519b55fb38900e0a808c3e3516b1f52c8b210c96968288def553962de0948"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
